### PR TITLE
[ RNN ] Implement initialization and forwarding of RNN Layer

### PR DESCRIPTION
**Pull request by jijoongmoon**
- [ RNN ] Add initialization of RNN

```
Add Initialization of RNN Layer
- Set dimension and Weight
- Add Unittest for initialization

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>
```

- [ RNN ] Implementaion of forwarding
```
- Implementation of RNN forwarding.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>
```
- [ RNN ] support various activation function

```
In this PR, various activation functions are supported using ActiFunc
Class.

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>
```

### CODE REVIEW

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -643,11 +643,17 @@ int NeuralNetwork::train_run() {
...
-      knn->net_hidden[0]->getGradientRef() = label;
+    layer_ = layer_node.layer;
+    if (layer_->getType() == "centroid_knn") {
+      layer_->net_hidden[0]->getGradientRef() = label;
+    }
+
+    if (layer_->getType() == "RNN"){
+      Tensor &var_ref= layer_->net_hidden[0]->getVariableRef();
+      // for now, we are using Xavir_normal for hidden initialization
+      var_ref.setRandNormal(0.0f, sqrtFloat(2.0f / (var_ref.width() + var_ref.height())));
```
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp]: 1. `HiddenLayer value` must be a weight for RNN, and not the output. If it is an output, it can be overwritten by other layers/derivatives, but that value is needed for the next iteration. Further, during inference, its value would be overwritten due to optimization but hidden value is needed for the next iteration.
2. Initialization of hidden value must be done inside the RNN layer itself, and not by the model. `centroid_knn` does it in the model which is wrong. It requires to pass the label for the training/validation and should be done by forwarding/backwarding like done for loss layer.

File: nntrainer/layers/rnn.cpp [hunk 2]
```diff
@@ -95,7 +96,38 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
...
+    weightAt(static_cast<int>(RNNParams::weight_xh)).getVariableRef();
+  Tensor &weight_hh =
+    weightAt(static_cast<int>(RNNParams::weight_hh)).getVariableRef();
+  Tensor &bias_h =
+    weightAt(static_cast<int>(RNNParams::bias_h)).getVariableRef();
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:110]: why use loops? Simply reshape to `batch * seq_length, 1, 1, input_dim[0].width()` and dot product with the weight.
Current code creates a lot of temporary memory allocations and is complicated making it prone to errors.

File: nntrainer/layers/rnn.cpp [hunk 3]
```diff
@@ -144,4 +144,9 @@ void RNNLayer::calcGradient() {
   // NYI
 }
 
+void RNNLayer::setActivation(ActivationType acti_type) {
+  Layer::setActivation(acti_type);
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:148]: please override `setProperty` and add this in `setProperty`.
`setActivation` will soon be removed.

File: nntrainer/layers/rnn.cpp [hunk 4]
```diff
@@ -95,7 +96,38 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
...
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor islice = input_.getBatchSlice(b, 1);
+    Tensor oslice = hidden_.getBatchSlice(b, 1);
+
+    for (unsigned int t = 0; t < islice.height(); ++t) {
+      Tensor xs = Tensor(TensorDim(1, 1, 1, islice.width()),
```
Comment (zhoonit) [inline nntrainer/layers/rnn.cpp:115]: Code clean: how about `TensorDim(islice.width())`?

File: nntrainer/layers/rnn.cpp [hunk 5]
```diff
@@ -95,7 +96,38 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
   }
 }
 
-void RNNLayer::forwarding(bool training) {}
+void RNNLayer::forwarding(bool training) {
```
Comment (zhoonit) [inline nntrainer/layers/rnn.cpp:99]: As far as I understand, RNN retains the (last) previous output and use it when inference mode. I wonder if this behavior is considered

File: nntrainer/layers/rnn.h [hunk 6]
```diff
@@ -93,10 +93,13 @@ class RNNLayer : public Layer {
   void setProperty(const PropertyType type,
                    const std::string &value = "") override;
 
+  void setActivation(ActivationType acti_type) override;
+
   static const std::string type;
 
 private:
   unsigned int unit;
+  ActiFunc acti_func;
```
Comment (kparichay) [inline nntrainer/layers/rnn.h:102]: Please set the default activation function to `TanH`. RNN defaults to TanH activation and a user specifying RNN without activation must still remain a valid RNN.

File: nntrainer/layers/rnn.cpp [hunk 7]
```diff
@@ -95,7 +96,38 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
...
+    Tensor oslice = hidden_.getBatchSlice(b, 1);
+
+    for (unsigned int t = 0; t < islice.height(); ++t) {
+      Tensor xs = Tensor(TensorDim(1, 1, 1, islice.width()),
+                         islice.getAddress(t * islice.width()));
+      unsigned int id = 0;
+      if (t > 0) {
+        id = t - 1;
+      }
+      Tensor hs = Tensor(TensorDim(1, 1, 1, oslice.width()),
```
Comment (zhoonit) [inline nntrainer/layers/rnn.cpp:121]: Tensor(Tensordim, float * data) will create a new copy so I think this will not update hidden.

Could you make a small `model` test to verify the output?

File: nntrainer/layers/rnn.cpp [hunk 8]
```diff
@@ -123,8 +123,8 @@ void RNNLayer::forwarding(bool training) {
       Tensor hs_prev = Tensor(TensorDim(1, 1, 1, oslice.width()),
                               oslice.getAddress(id * oslice.width()));
       // Calculate hs_t = tanh(Whh*h_(t-1) + Wxh*X_t))
-      hs =
-        xs.dot(weight_xh).add(hs_prev.dot(weight_hh).add(bias_h)).apply(tanh);
+      hs = xs.dot(weight_xh).add(hs_prev.dot(weight_hh).add(bias_h));
+      acti_func.run_fn(hs, hs);
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:127]: Please add a check in the graph that the next layer of the RNN must not be an activaiton layer.

File: nntrainer/graph/network_graph.cpp [hunk 9]
```diff
@@ -227,6 +228,12 @@ int NetworkGraph::realizeActivationType(Layer &current) {
 
   ActivationType act = current.getActivationType();
 
+  if (current.getType() == RNNLayer::type) {
+    // No need to add activation layer for RNN Layer
+    current.setActivation(act);
```
Comment (kparichay) [inline nntrainer/graph/network_graph.cpp:233]: `none` activation layer is also supported?

Comment (jijoongmoon) [reply nntrainer/graph/network_graph.cpp:233]: I'm thinking of. but later PR, default activation is tanh.

File: nntrainer/layers/rnn.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:99]: could you point out more specific? previous output??

File: nntrainer/graph/network_graph.cpp (hunk 9 above)
Comment (kparichay) [reply nntrainer/graph/network_graph.cpp:233]: I think `none` must not be supported. Can be opened later if requested.

File: nntrainer/layers/rnn.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:110]: I'm not thinking about any optimization here. at least this PR. We do not have any feature about the RNN. It will update one by one.

File: nntrainer/layers/rnn.cpp (hunk 4 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:115]: done

File: nntrainer/layers/rnn.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:99]: We discuss offline. I will update.

File: nntrainer/layers/rnn.h (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.h:102]: come up this next PR

File: nntrainer/layers/rnn.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:110]: The current code is quite difficult to understand. This can be updated to be written functionally within 3-4 lines which are quite easy to reason about and review.

Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:110]: Ok. I will try to make it easier then. Besides, I think loop in batch is necessary because of update the hidden vector between batch.

File: nntrainer/layers/rnn.cpp (hunk 7 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:121]: Right. I will update this.

File: nntrainer/layers/rnn.cpp (hunk 8 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:127]: In network graph, there is realizeActivationLayer to create additional activation layer. but for RNN Layer, there is no need to do it.. so.. it just set default tanh if type is ACT_NONE. Talking about the user intended addition activation layer.. I think we don't need to care about. That's user intension. Also to calculate softmax.. user also have responsibility to add Fully Connected Layer. I think the restriction of activation layer is totally depending on user.

File: nntrainer/layers/rnn.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:148]: I will. Thanks.

File: nntrainer/layers/rnn.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:110]: You can refer to the FC layer implementation or use FC layer inside RNN as well :)
The operation performed here is exactly the same as FC layer with an extra dimension (which can be merged with batchsize for simplicity).

File: nntrainer/layers/rnn.h (hunk 6 above)
Comment (kparichay) [reply nntrainer/layers/rnn.h:102]: Great.

File: nntrainer/layers/rnn.cpp (hunk 8 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:127]: The issue is that RNN uses activation function internally which is being operated in-place in the current code.
So, if an activation layer is added after it, which is also optimized to work in-place, it will create issues.
If there is no such restriction of activation layer being added to the graph, please add a `TODO:` in `inPlaceOptimize` in the graph to ensure that if the input layer of activation is RNN or RNN has activation internally (to be more precise), then that activation must not be inplace. This will later be updated to each layer provided `isOperatingInPlace` with #867.

File: nntrainer/layers/rnn.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:120]: > You can refer to the FC layer implementation or use FC layer inside RNN as well :)
> The operation performed here is exactly the same as FC layer with an extra dimension (which can be merged with batchsize for simplicity).

I think there is more than that. This is updated.

File: nntrainer/layers/rnn.cpp (hunk 8 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:165]: Let's see. how it works. RNNLayer still in the middle of development. so this kind of optimization can be considered later.

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/models/neuralnet.cpp]: right. it removed. Thanks.

File: nntrainer/layers/rnn.cpp (hunk 8 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp:165]: Alright. Please add a TODO as comments to keep it as a reminder.

File: nntrainer/layers/rnn.cpp [hunk 10]
```diff
@@ -54,7 +105,66 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
...
+  Tensor temp;
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
+    Tensor islice = input_.getBatchSlice(b, 1);
+    Tensor oslice = hidden_.getBatchSlice(b, 1);
+    float *hidden_next = oslice.getData();
+
+    if (!training) {
+      if (b > 0) {
+        float *hidden_pre =
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:127]: I am a little confused. I think the hidden layer is not shared across the batch. Each batch has its own hidden information in a generic RNN.

Comment (kparichay) [reply nntrainer/layers/rnn.cpp:127]:
```python
from tensorflow.keras.layers import Input, SimpleRNN
in = Input(shape=(3, 1), batch_size=10)
print(in.shape) # TensorShape([10, 3, 1])
rnn = SimpleRNN(5)(in) # can also enable return_state=True to get the state
print(rnn.shape) # TensorShape([10, 5]) - hidden cell value
```
The hidden state between batch size should not share their values.

Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:127]: for the inference, I guess it should. but backwarding, it's not.

Comment (kparichay) [reply nntrainer/layers/rnn.cpp:127]: IMHO, even for inference, it should be not.

Comment (zhoonit) [reply nntrainer/layers/rnn.cpp]: There are number of points that needs clarification.

I think there are two ways.

1. Keep states for the batch only 🌴 
2. Presist states 🐟 

The one I am referencing suggested to train with 🌴 and inference 🐟, Jijoong told me the implementation is done in 🐟  but later add `stateful=false` as a property.

keras seems to have two options.
https://keras.io/guides/working_with_rnns/#crossbatch-statefulness

File: nntrainer/models/neuralnet.cpp [hunk 11]
```diff
@@ -632,10 +632,11 @@ int NeuralNetwork::train_run() {
       layer_->net_hidden[0]->getGradientRef() = label;
     }
 
-    if (layer_->getType() == "RNN"){
-      Tensor &var_ref= layer_->net_hidden[0]->getVariableRef();
+    if (layer_->getType() == "RNN") {
+      Tensor &var_ref = layer_->net_hidden[0]->getVariableRef();
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:636]: I wonder if we can handle this on `RNN::initialize()`.

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/layers/rnn.cpp (hunk 10 above)
Comment (kparichay) [reply nntrainer/layers/rnn.cpp]: Stateful and batch are very different concepts.
Stateful talks about keeping the hidden state across different runs.
Batch is about different data points running in parallel.

The behavior of RNN remains the same in inference and train for the batch and stateful. NNTrainer's RNN's default behavior must match other frameworks default behavior to avoid confusion.

> https://keras.io/guides/working_with_rnns/#crossbatch-statefulness

Although the name is `crossbatch`, the state is independent across the batch. This is just a way to selectively reset the state when needed. This does not shares the state across the batch (multiple data points given at once). The `batch` terminology used in the article refers to `iteration`, and it is equivalent to a simple behavior, where we can relinquish the power to reset the hidden state to the user.

File: nntrainer/models/neuralnet.cpp [hunk 12]
```diff
@@ -625,11 +625,18 @@ int NeuralNetwork::train_run() {
...
-      knn->net_hidden[0]->getGradientRef() = label;
+    layer_ = layer_node.layer;
+    if (layer_->getType() == "centroid_knn") {
+      layer_->net_hidden[0]->getGradientRef() = label;
+    }
+
+    if (layer_->getType() == "RNN") {
+      Tensor &var_ref = layer_->net_hidden[0]->getVariableRef();
+      // for now, we are using Xavir_normal for hidden initialization
+      var_ref.setRandNormal(
```
Comment (kparichay) [inline nntrainer/models/neuralnet.cpp:638]: is this even required anymore?

File: nntrainer/layers/rnn.cpp [hunk 13]
```diff
@@ -54,7 +109,49 @@ void RNNLayer::setProperty(const PropertyType type, const std::string &value) {
...
+    weightAt(static_cast<int>(RNNParams::bias_h)).getVariableRef();
+
+  Tensor &hidden_ = net_hidden[0]->getVariableRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  Tensor temp;
+  Tensor hs_prev;
+  Tensor hs;
+
+  for (unsigned int b = 0; b < input_dim[0].batch(); ++b) {
```
Comment (kparichay) [inline nntrainer/layers/rnn.cpp:126]: Please add a TODO here to try and transpose by the batch dimension and see if its faster that this implementation.

File: nntrainer/models/neuralnet.cpp (hunk 11 above)
Comment (jijoongmoon) [reply nntrainer/models/neuralnet.cpp:636]: removed

File: nntrainer/models/neuralnet.cpp (hunk 12 above)
Comment (jijoongmoon) [reply nntrainer/models/neuralnet.cpp:638]: removed, thanks.

File: nntrainer/layers/rnn.cpp (hunk 13 above)
Comment (jijoongmoon) [reply nntrainer/layers/rnn.cpp:127]: done.

**Review by kparichay — APPROVED** [review]
LGTM
Please remember that this version does not reset the hidden state per iteration, and behaves like a stateful RNN.
